Add Installer Script for upcoming PHAR releases

alexfornuto opened the pull request:
## Summary of Changes

 - Adds a new installer script, to be used as a one-line installation method for Terminus .phar files.
 - Updates the README with instructions on how to use it.
 - Updates the CircleCI configuration to test the installer script.

## Remaining Work

- [ ] Point `$package` to `terminus` once a release phar is in place.
- [ ] Improve the one-liner provided in the README to avoid leaving the install script in the working directory.
- [ ] Include more checks in the script for things like existing instances of terminus.
- [ ] Squash the commits, hiding the shameful mistakes I made.

Commit:
Add installer.php to install upcoming phar releases

greg-1-anderson commented:
One thing I forgot to mention -- the URL for the phar will always have a version number in it; GitHub does not have a "latest release" alias URL.

Example:
- [Releases page](https://github.com/pantheon-systems/updatinate/releases)
- [phar url](https://github.com/pantheon-systems/updatinate/releases/download/0.5.2/updatinate.phar)

Your script could download a well-known version, and then use `terminus self:update` to update it to the latest version. That would be easy, but it would involve downloading Terminus twice.

Examples of how to use the GitHub API to find the latest release can be found in the [self update repository](https://github.com/consolidation/self-update/blob/master/src/SelfUpdateCommand.php#L65:L73).

alexfornuto commented:
@greg-1-anderson I think the answer there is that the installer should always download the latest stable release, and part of the release process should be updating the version the installer downloads.

One way to do that is with a .version file, which would be trivial to overwrite during release, and could be read by the installer script. Would that solution work?

P.S. That also means I should probably include a check for an existing version of terminus, and fail if found telling them to either update or remove existing versions.

greg-1-anderson commented on installer.php:
755 probably better

greg-1-anderson commented:
To check standard coding conventions on this file, open up composer.json and change:
```
      "cs": "phpcs --standard=PSR2 --severity=1 -n tests/unit_tests bin/terminus src",
```
to:
```
      "cs": "phpcs --standard=PSR2 --severity=1 -n tests/unit_tests bin/terminus src installer.php",
```
Then run `composer cs` to check.

Commit:
Muliple updates:

 - include .profile as a possible conf file
 - define a varable to track the updating of conf files
 - improve the syntax by which conf files are written and errors caught
 - alerts the user if no conf files were updated

Commit:
complete refactor

greg-1-anderson commented on installer.php:
```diff
@@ -0,0 +1,105 @@
+// Define environemnt variables.
+$pathUpdated = false;
+$paths = explode(":", getenv('PATH'));             // Creates an array with all paths in $PATH
+$installdir = (getenv('HOME') . "/.terminus/bin"); // Creates a string with the desired installation path
```
I know I suggested this path when we were chatting, but I think it was a mistake. ~/.terminus is the location that Terminus is already using to store configuration files, and we shouldn't put the binary in the same path as the config.

I think that `$HOME/bin` is a pretty good default.

alexfornuto replied:
My concern remains that inexperienced users will see `bin` in finder and delete it. What about `$HOME/.bin` as a middle-ground?

alexfornuto replied:
OR, I could try go work it to use `/usr/local/bin`, and prompt for sudo if required.

alexfornuto replied:
In fact, that's exactly what I did in https://github.com/pantheon-systems/terminus/pull/1927/commits/7dabceeae53fd9eb227bf93acab7bed0cc7cdd62

alexfornuto replied:
Bonus points: since `/usr/local/bin` is almost certainly already in `$PATH`, we can be less concerned about writing to a shell conf file.

greg-1-anderson commented on installer.php:
```diff
@@ -0,0 +1,105 @@
+function checkpath($paths, $installdir)
+{
+
+    if (in_array($installdir, $paths)) {
```
Optimization:
```
return in_array($installdir, $paths);
```

alexfornuto replied:
Optimized, Thanks!

greg-1-anderson commented on installer.php:
```diff
@@ -0,0 +1,105 @@
+    echo("Downloaded to " . $installdir . "\n\n");
+} else {
+    echo("Download unsuccessful.\n\n");
+    exit();
```
`exit(1);` for failures

greg-1-anderson replied:
... or use `exit([string])` like elsewhere, for consistency.

alexfornuto replied:
Made consistent, TY

greg-1-anderson commented on installer.php:
```diff
@@ -0,0 +1,105 @@
+// If ~/.terminus/bin isn't in path, add it.
+if (checkpath($paths, $installdir) === false) {
+    foreach ($rcfiles as $rcfile) {
+        if (file_exists(getenv('HOME') . "/$rcfile") && is_writable(getenv('HOME') . "/$rcfile")) {
```
If multiple rc files exist, this code will update the PATH in all of them. It would probably be better to just pick one.

alexfornuto replied:
If multiple rc files exist, it's because there are multiple shells on the system. None of the rc files listed are both used by default by any one shell. By updating any it finds, we ensure that Terminus remains in the path even if the user, for example, ditches fish and goes back to ZSH or bash.

Commit:
optimizations suggested by @greg-1-anderson

Commit:
Install to /usr/local/bin, prompt for sudo if needed

Commit:
cleanups for previous commit

ataylorme commented on README.md:
```diff
@@ -32,6 +32,15 @@ Once you have at least the requirements installed, you can install Terminus via
+
+### Direct Installation
+
+Copy the following one-liner to your terminal:
```
Should we mention that PHP (perhaps a minimum version) is required?

alexfornuto replied:
We do, [above](https://github.com/pantheon-systems/terminus/#required).

ataylorme commented:
A little late to the version conversation but you can [get the latest release from the GitHub API](https://developer.github.com/v3/repos/releases/#get-the-latest-release)

I tested `curl https://api.github.com/repos/pantheon-systems/terminus/releases/latest` and it worked for me. API rate limits shouldn't be a worry running a single call.

alexfornuto commented:
@ataylorme this script uses the API to pull the latest release. Right now it's configured with `$package = updatinate` because no releases of Terminus yet have a phar file to pull from the API output.

**P.S.**: To clarify, the path I'm using isn't `latest`, but will use whatever version is the latest, and by doing so I have that version number to use in my output.

Commit:
typo

SaraMcCutcheon reviewed (commented):
@alexfornuto I am concerned that by adding an installer script like this we open ourselves up to many of the pitfalls of my own installer endeavor - by taking responsibility for the installation of Terminus we take it upon ourselves to fix any issues that might arise in that installation. Do many of our users find they have trouble knowing where to install PHARs? This script also seems like it would be terribly difficult to write tests for.

alexfornuto commented:
>  ourselves up to many of the pitfalls of my own installer endeavor

This is not a complicated piece of software, it's a single file that does a single task.

>  by taking responsibility for the installation of Terminus we take it upon ourselves to fix any issues that might arise in that installation

IMHO not an unreasonable burden for software maintainers to maintain proper installation methods.

> Do many of our users find they have trouble knowing where to install PHARs? 

I think @pantheon-systems/cse can confirm that there are plenty of users who don't already have a go-to place to put non-system-maintained executable files, who would benefit from a one-liner solution.

> This script also seems like it would be terribly difficult to write tests for.

The tests could be part of CircleCI, written in path, and as easy as:

```
curl  `path-to-installer` && php installer.php
terminus --version
```

In other words, the only test it needs is if it's installed or not. Since CircleCI images have passwordless sudo, we could easily run tests for root and non-root users to confirm the sudo request works.

greg-1-anderson commented:
I share some of @TeslaDethray 's concerns. Downloading a phar from GitHub is not hard, and maintenance of an installer is additional work we would need to take on that we simply wouldn't need to do if we solved this in documentation. We could consider looking at the Composer installer as one way to maybe reduce maintenance needs, if it could be adapted to our uses.

That said, I also have words in favor of an installer. I think that the main problem that users had with the terminus-installer was that it left it up to the user to decide where to install terminus. This left users vulnerable to some pretty difficult problems that could crop up if they happened to install other Composer applications in the same location. While a more opinionated installer might have mitigated this, the primary problem stems from the way a Composer-based install works. Users would have had these same problems if they installed Terminus by running Composer directly (and some did). A phar distribution will not have these problems, whether it is manually installed or copied from an installer.

I also think that simple tests would be easy to write and maintain.

Overall, I am +1 on having an installer, but I care 2 out of 10. Solving this in docs is a viable strategy. I do appreciate the hard work @alexfornuto has done here to try to make things a bit easier on our end users.

greg-1-anderson commented:
Here's a three-line installation alternative to an installer:
```
wget https://github.com/pantheon-systems/terminus/releases/download/2.0.0/terminus.phar
chmod +x terminus.phar 
 ./terminus.phar self:update
```
Note that we invariantly download the first 2.0 release of Terminus and then use its built-in self:update command to bring it up to the most recent available release.  Follow-on instructions to set up a $PATH or make a bash alias or rename and move terminus.phar to terminus may also be necessary, but for folks nominally familiar with such things, the three lines above should be sufficient.

alexfornuto commented:
@greg-1-anderson that doesn't appear to work for fresh installations:

```
$ php terminus.phar self:update
PHP Fatal error:  Uncaught ReflectionException: Class DOMDocument does not exist in phar:///home/alex/terminus.phar/vendor/consolidation/output-formatters/src/Transformations/DomToArraySimplifier.php:24
Stack trace:
#0 phar:///home/alex/terminus.phar/vendor/consolidation/output-formatters/src/Transformations/DomToArraySimplifier.php(24): ReflectionClass->isSubclassOf('DOMDocument')
#1 phar:///home/alex/terminus.phar/vendor/consolidation/output-formatters/src/FormatterManager.php(371): Consolidation\OutputFormatters\Transformations\DomToArraySimplifier->canSimplify(Object(ReflectionClass))
#2 phar:///home/alex/terminus.phar/vendor/consolidation/output-formatters/src/FormatterManager.php(193): Consolidation\OutputFormatters\FormatterManager->canSimplifyToArray(Object(ReflectionClass))
#3 phar:///home/alex/terminus.phar/vendor/consolidation/output-formatters/src/FormatterManager.php(188): Consolidation\OutputFormatters\FormatterManager->isValidDataType(Object(Consolidation\OutputFormatters\Formatters\NoOutputFormatter), Object(Re in phar:///home/alex/terminus.phar/vendor/consolidation/output-formatters/src/Transformations/DomToArraySimplifier.php on line 24
```

I've moved this script to a new [third-party repo](https://github.com/alexfornuto/terminus-installer) to continue development.

greg-1-anderson commented:
You need the php-xml extension. You will get similar errors if you try to do Drupal development without that installed.  This extension comes under different names; php-xml, php-dom, php-doc. See https://stackoverflow.com/questions/14395239/class-domdocument-not-found

alexfornuto commented:
For those of you following along at home, `php-xml` is assumed but not listed as a dependency in `README.md`. Update staged in #1959.

alexfornuto commented:
>Follow-on instructions to set up a $PATH or make a bash alias or rename and move terminus.phar to terminus may also be necessary, but for folks nominally familiar with such things, the three lines above should be sufficient.

My script isn't for those who know about $PATH, Aliases, etc. It's for the web developer who's never had to touch the CLI before, for whom Terminus is their first command-line tool.

If Pantheon wants to be the platform where people learn to be web developers, then we can't expect them to already be developers. My script doesn't assume knowledge that you may take for granted because it's been so long since you didn't know it.

My script avoids frustration for people who may want to learn about all that stuff, but not right now while they're in the middle of setting up their Pantheon workflow.

I moved the script to its own repository so that I can continue to work on it independently.
